### PR TITLE
placement no longer gets cloned for every shard in MarkShardAvaialble

### PR DESCRIPTION
**Pull request by dgromov**
The placement clone is now create once before going through all the
shards. This commit also moves a bunch of stuff into placement.go from
the algo module.

### CODE REVIEW

File: services/placement/placement.go [hunk 1]
```diff
@@ -316,3 +354,79 @@ func (s ByIDAscending) Less(i, j int) bool {
...
+	for i, instance := range list {
+		if instance.ID() == instanceID {
+			return append(list[:i], list[i+1:]...)
+		}
+	}
+	return list
+}
+
+// MarkShardAvailable marks the state of a shard to available
+func MarkShardAvailable(p services.ServicePlacement, instanceID string, shardID uint32) (services.ServicePlacement, error) {
```
Comment (cw9) [inline services/placement/placement.go]: nit: you could create a markShardAvailable which does everything MarkShardAvailable does now, and MarkShardAvailable just do a clone and then call markShardAvailable, so the markShardAvailable can be reused in MarkAllShardsAsAvailable as well

File: services/placement/algo/sharded_helper_test.go [hunk 2]
```diff
@@ -191,54 +191,14 @@ func TestNonLeavingInstances(t *testing.T) {
...
-	_, err = MarkShardAvailable(p, "i2", 3)
-	assert.Error(t, err)
-	assert.Contains(t, err.Error(), "does not exist in source instance")
-
-	_, err = MarkShardAvailable(p, "i2", 2)
-	assert.Error(t, err)
-	assert.Contains(t, err.Error(), "not leaving instance")
-}
-
 func TestRemoveInstanceFromArray(t *testing.T) {
```
Comment (cw9) [inline services/placement/algo/sharded_helper_test.go]: this test should be moved

Comment (dgromov) [inline services/placement/algo/sharded_helper_test.go]: I did. Its in placement_test now

File: services/placement/algo/sharded_helper_test.go [hunk 3]
```diff
@@ -191,54 +191,14 @@ func TestNonLeavingInstances(t *testing.T) {
...
 		placement.NewEmptyInstance("i2", "", "", "endpoint", 1),
 	}
 
-	assert.Equal(t, instances, removeInstance(instances, "not_exist"))
-	assert.Equal(t, []services.PlacementInstance{placement.NewEmptyInstance("i2", "", "", "endpoint", 1)}, removeInstance(instances, "i1"))
+	assert.Equal(t, instances, placement.RemoveInstance(instances, "not_exist"))
+	assert.Equal(t, []services.PlacementInstance{placement.NewEmptyInstance("i2", "", "", "endpoint", 1)}, placement.RemoveInstance(instances, "i1"))
 }
 
 func TestCopy(t *testing.T) {
```
Comment (cw9) [inline services/placement/algo/sharded_helper_test.go]: this should be moved as well

**Pull request merged**